curvine-ufs/src/oss_hdfs/oss_hdfs_reader.rs

```diff
@@ -0,0 +1,568 @@
+            if start_status != JindoStatus::Ok {
+                unsafe {
+                    drop(Box::from_raw(
+                        userdata as *mut oneshot::Sender<(JindoStatus, i64, Option<String>)>,
+                    ))
+                };
+                return Ok(self.length);
+            }
+        }
+
+        let (status, length, _err) = match rx.await {
+            Ok(v) => v,
+            Err(_) => return Ok(self.length),
+        };
+        if status != JindoStatus::Ok {
+            return Ok(self.length);
+        }
+        Ok(length)
+    }
+}
+
+/// OSS-HDFS Reader implementation using JindoSDK C++ library via FFI
+pub struct OssHdfsReader {
+    pub(crate) reader_handle: Arc<Mutex<Option<JindoReaderHandle>>>,
+    pub(crate) path: Path,
+    pub(crate) length: i64,
+    pub(crate) pos: i64,
+    pub(crate) chunk_size: usize,
+    pub(crate) status: FileStatus,
+    pub(crate) chunk: DataSlice,
+}
+
+// Safety: JindoReaderHandle is a *mut c_void pointer from FFI.
+// The handle is protected by Mutex, and all FFI calls are thread-safe.
+// The underlying C++ library handles thread safety internally.
+unsafe impl Send for OssHdfsReader {}
+unsafe impl Sync for OssHdfsReader {}
+
+impl Reader for OssHdfsReader {
+    fn status(&self) -> &FileStatus {
+        &self.status
+    }
+
+    fn path(&self) -> &Path {
+        &self.path
+    }
+
+    fn len(&self) -> i64 {
+        self.length
+    }
+
+    fn chunk_mut(&mut self) -> &mut DataSlice {
+        // Return a reference to the actual chunk buffer
+        // This buffer is used by the Reader trait's default implementations
+        // (read_chunk, etc.) but we override read_chunk0 to read directly from JindoSDK
+        &mut self.chunk
+    }
+
+    fn chunk_size(&self) -> usize {
+        self.chunk_size
+    }
+
+    fn pos(&self) -> i64 {
+        self.pos
+    }
+
+    fn pos_mut(&mut self) -> &mut i64 {
+        &mut self.pos
+    }
+
+    async fn read_chunk0(&mut self) -> FsResult<DataSlice> {
+        // If file is empty or we've reached the end, return an empty buffer (keeps capacity reusable).
+        if self.length == 0 || self.pos >= self.length {
+            // Keep the internal chunk as an empty Buffer so future reads can reuse capacity.
+            let prev = std::mem::replace(&mut self.chunk, DataSlice::Empty);
+            let mut buf = match prev {
+                DataSlice::Buffer(b) => b,
+                _ => BytesMut::new(),
+            };
+            buf.clear();
+            return Ok(DataSlice::Buffer(buf));
+        }
+
+        // Only read up to remaining bytes.
+        let remaining = (self.length - self.pos).max(0) as usize;
+        let want = remaining.min(self.chunk_size());
+
+        // Reuse internal buffer stored in `self.chunk` by the shared Reader impl.
+        let prev = std::mem::replace(&mut self.chunk, DataSlice::Empty);
+        let mut buffer = match prev {
```

Comment:
How to obtain a buffer:
  self.buf.reserve(len as usize);
        unsafe {
            self.buf.set_len(len as usize);
        }
        let mut buf = self.buf.split_to(len as usize);